### ISSUE TITLE
Already existing protocols/solutions

### ISSUE THREAD

**Issue by bilderbuchi** [pymeasure/experiment_control_protocol#2]
From [here](https://github.com/pymeasure/pymeasure/discussions/676#discussioncomment-4317207)

We should really check what protocols have already been defined. Protocol design is hard and finicky work, and it would pay to make use of the work of *those who came before*. I got the feeling that we can't be the first ones concerned with this problem space. 

If there is something that
* we could plug in to
* we could take a subpart and use that (e.g. because we don't need all the complexity)
* we could take as a template and adapt to our needs (I guess ZMQ majordomo falls into this)

we should take a l_ong hard look_ at **reusing** instead of coming up with our own stuff.

Specifically,
* What about the big distributed things like [Taurus](https://taurus-scada.org/index.html), [Tango](https://www.tango-controls.org/) or [Sardana](https://www.sardana-controls.org/) (maybe not EPICS, as that is very old afaik)
*  I think the [yaq](https://yaq.fyi/) folks also spent a lot of thought on this -- maybe @untzag can advise here?
* Probably a strong contender: [Bluesky](https://nsls-ii.github.io/bluesky/). I remember bluesky making the strongest impression on me back when I wrote my own tool referenced upthread.
  * Skimming the docs makes me think that these folks have the right ideas, and there might be enough commonality to make use of. Can we maybe plug ourselves in like [yaq seems to do](https://github.com/bluesky/yaqc-bluesky)?
  * They have understandable documentation on the [message protocol](https://nsls-ii.github.io/bluesky/msg.html), [data model](https://blueskyproject.io/event-model/data-model.html), [HW interface](https://nsls-ii.github.io/bluesky/hardware.html)
* Any others?

**Comment by bilderbuchi** [comment]
What's missing is a low entry point, many of those solutions, while powerful, are quite complicated to set up.
Bluesky would be preferable to Sardana, but those synchrotron-focused projects are not that suitable for "quickly getting started" (with a small setup) or ramping up on if unfamiliar.
Bluesky seems to focus on EPICS/Ophyd, possibly yaq is working on getting _other_ things outside those environments to integrate.

It seems hard to get "just" the communication parts, without involving RunEngines (BlueSky), daemons (yaq), device servers (Tango), etc.

@bmoneke failed to install a working Tango installation. Tango needs C++ skills and SQL databases.

**Comment by bmoneke** [comment]
One of the projects (I do not know which one) sends "message" objects (bluesky does, but I am looking for another one) and they are queued in the recipient.

**Comment by bilderbuchi** [comment]
There's an interesting paper "How to Design &amp; Implement a Modern Communication Middleware Based on ZeroMQ" ([https://doi.org/10.18429/JACoW-ICALEPCS2017-MOBPL05](http://jacow.org/icalepcs2017/papers/mobpl05.pdf)). I think it's definitely relevant to our design effort, and only a couple of pages. 
One decision in the paper that we probably will decide differently (because our trade-offs are different from CERN's) is to write a serialiser themselves instead of using something like Apache Avro.

**Comment by bmoneke** [comment]
Thanks for that source, it is really insightful.

**Comment by bmoneke** [comment]
A few ideas from that [paper](https://accelconf.web.cern.ch/icalepcs2017/papers/mobpl05.pdf), which might be good for us or spark a discussion:

- Separation of the transport layer and business layer (content) is a good idea.
- They used test driven development, which helped them a lot.
- They dropped a publishing scheme in favour of simplicity and it did not fit their requirement (getting to know some initial information). They are right, that you do not save on bandwidth (a published message is sent as individual messages to the recipients on the tcp backbone). It might save some programming work (a Component does not have to manage the subscribers of changes manually, as zmq does that for us). For a "data stream without beginning and end" it is great, however.
- They wrote their own serialization scheme to reduce dependencies and increase performance. (I prefer to start with a ready done serialization scheme, if possible).
- Each Component has a thread dedicated for message handling.
- Each message counts as a heartbeat 
- DEALER/ROUTER socket pairs.

**Comment by bilderbuchi** [comment]
Yeah, this would ~ also have been my summary of the main points. Thank you for the excerpt!

**Comment by bmoneke** [comment]
Here is the COAP definition (an IoT protocol), which can serve as a language / definition reference and hints for protocol ideas:
https://www.rfc-editor.org/rfc/rfc7252

wikipedia short article: https://en.wikipedia.org/wiki/Constrained_Application_Protocol

**Comment by VigneshVSV** [comment]
My package's slowly upcoming doc is here : https://hololinked.readthedocs.io/en/latest/index.html

**Comment by VigneshVSV** [comment]
Just an update - If you guys are also doing RPC, I simply propose a merger.

**Comment by VigneshVSV** [comment]
I dont exactly see a double-effort as a negative thing because one still learns a lot and its better to own one's knowledge thoroughly with practice.   

It would be best if we once discuss in terms of specifications of user features to see if a merger is possible because both are based on ZeroMQ.  May be we can share a power point presentation to each other?

Can I write to your email ?`

**Comment by bilderbuchi** [comment]
> It would be best if we once discuss in terms of specifications of user features to see if a merger is possible because both are based on ZeroMQ. May be we can share a power point presentation to each other?

You can already study what we have come up with on the specification side at the rendered documentation: https://leco-laboratory-experiment-control-protocol.readthedocs.io/en/latest/ :-)

**Comment by VigneshVSV** [comment]
For future reference - moved conversation to email.
